pkg/catalog/manager/traverse.go

```diff
@@ -4,23 +4,41 @@ import (
 	"fmt"
 	"strings"
 
+	"github.com/pkg/errors"
 	"github.com/rancher/rancher/pkg/catalog/helm"
 	"github.com/rancher/types/apis/management.cattle.io/v3"
 	"github.com/sirupsen/logrus"
 	"gopkg.in/yaml.v2"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 )
 
-func traverseFiles(repoPath string, catalog *v3.Catalog) ([]v3.Template, []string, map[string]v3.VersionCommits, []error, error) {
+func (m *Manager) traverseAndUpdate(repoPath, commit string, catalog *v3.Catalog) error {
 	logrus.Info("traversing files and generating templates")
 	index, err := helm.LoadIndex(repoPath)
 	if err != nil {
-		return nil, nil, nil, nil, err
+		return err
+	}
+
+	// list all existing templates
+	templateMap, err := m.getTemplateMap(catalog.Name)
+	if err != nil {
+		return err
+	}
+	// list all templateContent tag
+	templateContentMap := map[string]struct{}{}
```

Comment:
nit: may be initialize map after doing the list, so in case of error we save on map creation call